### PR TITLE
Process data request messages

### PR DESCRIPTION
**Pull request by vitbaq**

### CODE REVIEW

File: src/parser.c [hunk 1]
```diff
@@ -478,55 +478,38 @@ struct l_queue *parser_mydevices_to_list(const char *json_str)
...
 
+		if (!l_queue_push_tail(list,
+				l_memdup(&sensor_id, sizeof(sensor_id))))
+			goto fail;
 	}
 
-	json_object_put(jobjarray);
+	return list;
 
-	if (l_queue_isempty(list)) {
```
Comment (ramonhpr) [inline src/parser.c]: If you remove this verification it can lead to a memory leak.

File: src/parser.c [hunk 2]
```diff
@@ -478,55 +478,38 @@ struct l_queue *parser_mydevices_to_list(const char *json_str)
...
+			goto fail;
 
-		/* Order matters: Add to tail to generate the same json */
-		l_queue_push_tail(list,
-				  l_memdup(&sensor_id, sizeof(sensor_id)));
+		sensor_id = json_object_get_int(jobjentry);
 
+		if (!l_queue_push_tail(list,
+				l_memdup(&sensor_id, sizeof(sensor_id))))
+			goto fail;
```
Comment (ramonhpr) [inline src/parser.c]: If in one of the interactions the function `l_memdup` be called you should also free the elements, no just the list.

File: test/mock-connector.py [hunk 3]
```diff
@@ -38,18 +30,14 @@
...
+    format='%(asctime)s %(message)s',
+    level=logging.INFO,
+    datefmt='%Y-%m-%d %H:%M:%S')
 
-channel.queue_bind(
-        exchange=cloud_exchange, queue=queue_name, routing_key='device.*')
-channel.queue_bind(
-        exchange=cloud_exchange, queue=queue_name, routing_key='data.*')
 
 def callback(ch, method, properties, body):
```
Comment (ramonhpr) [inline test/mock-connector.py]: Since we have more than one callback. Rename it

File: src/msg.c [hunk 4]
```diff
@@ -1400,7 +1400,7 @@ static void start_timeout(struct l_timeout *timeout, void *user_data)
 	l_timeout_modify(timeout, 5);
 }
 
-static void append_queue(void *data, void *user_data)
+static void append_update_queue(void *data, void *user_data)
```
Comment (alla3) [inline src/msg.c]: The name of this function is not clear enough. Please, update it.

Comment (vitbaq) [reply src/msg.c]: For me, it is clear enough. What do you suggest?

Comment (vitbaq) [reply src/msg.c]: i change to append_on_update_list

File: test/mock-connector.py [hunk 5]
```diff
@@ -63,8 +51,49 @@ def callback(ch, method, properties, body):
...
+def msg_request(args):
+    channel.basic_publish(exchange=fog_exchange,
+                          routing_key=KEY_REQUEST, body=args.message)
+
+parser = argparse.ArgumentParser(description='Mock KNoT Fog Connector')
+parser.set_defaults(func=consume)
+subparsers = parser.add_subparsers(help='sub-command help', dest='subcommand')
+
+parser_update = subparsers.add_parser('send-update', help='a help')
+parser_update.add_argument('message', type=str, help='message help')
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py [hunk 6]
```diff
@@ -63,8 +51,49 @@ def callback(ch, method, properties, body):
...
+parser_update.set_defaults(func=msg_update)
+
+parser_request = subparsers.add_parser('send-request', help='a help')
+parser_request.add_argument('message', type=str, help='message help')
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py [hunk 7]
```diff
@@ -63,8 +51,49 @@ def callback(ch, method, properties, body):
...
+parser_request.set_defaults(func=msg_request)
+
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py [hunk 8]
```diff
@@ -63,8 +51,49 @@ def callback(ch, method, properties, body):
...
+                          routing_key=KEY_UPDATE, body=args.message)
+def msg_request(args):
+    channel.basic_publish(exchange=fog_exchange,
+                          routing_key=KEY_REQUEST, body=args.message)
+
+parser = argparse.ArgumentParser(description='Mock KNoT Fog Connector')
+parser.set_defaults(func=consume)
+subparsers = parser.add_subparsers(help='sub-command help', dest='subcommand')
+
+parser_update = subparsers.add_parser('send-update', help='a help')
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py [hunk 9]
```diff
@@ -63,8 +51,49 @@ def callback(ch, method, properties, body):
...
+parser_update.add_argument('message', type=str, help='message help')
+parser_update.set_defaults(func=msg_update)
+
+parser_request = subparsers.add_parser('send-request', help='a help')
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py (hunk 3 above)
Comment (vitbaq) [reply test/mock-connector.py]: done

File: test/mock-connector.py [hunk 10]
```diff
@@ -38,6 +39,27 @@
...
+            "data": [{"sensor_id": 253, "data": True}]
+        }
+    return json.dumps(msg)
+
+def __parse_request_message(msg_file):
+    if msg_file:
+        with open(msg_file) as fd:
+            msg = json.load(fd)
+    else:
+        msg = {
```
Comment (ramonhpr) [inline test/mock-connector.py]: Create a global variable `defaultMsgRequest`

File: test/mock-connector.py [hunk 11]
```diff
@@ -38,6 +39,27 @@
     level=logging.INFO,
     datefmt='%Y-%m-%d %H:%M:%S')
 
+def __parse_update_message(msg_file):
+    if msg_file:
+        with open(msg_file) as fd:
+            msg = json.load(fd)
+    else:
+        msg = {
```
Comment (ramonhpr) [inline test/mock-connector.py]: Create a global variable `defaultMsgUpdate`

File: test/mock-connector.py [hunk 12]
```diff
@@ -59,27 +81,34 @@ def msg_consume(args):
     channel.start_consuming()
 
 def msg_update(args):
+    msg = __parse_update_message(args.json_msg_file)
```
Comment (ramonhpr) [inline test/mock-connector.py]:
```
with open(msg_file) as fd:
  channel.basic_publish(exchange=fog_exchange,
                          routing_key=KEY_UPDATE, body=fd.read())
```

File: test/mock-connector.py [hunk 13]
```diff
@@ -59,27 +81,34 @@ def msg_consume(args):
...
     channel.basic_publish(exchange=fog_exchange,
-                          routing_key=KEY_UPDATE, body=args.message)
+                          routing_key=KEY_UPDATE, body=msg)
 def msg_request(args):
+    msg = __parse_request_message(args.json_msg_file)
```
Comment (ramonhpr) [inline test/mock-connector.py]:
```
with open(msg_file) as fd:
  channel.basic_publish(exchange=fog_exchange,
                          routing_key=KEY_REQUEST, body=fd.read())
```

File: test/mock-connector.py [hunk 14]
```diff
@@ -59,27 +81,34 @@ def msg_consume(args):
...
 parser_update = subparsers.add_parser('send-update', help='Sends a message to \
     update the sensor in device')
-parser_update.add_argument('message', type=str, help='Update message to be \
-    sent. Format: {"id": <device_id>, "data":[{"sensor_id": <sensor_id>, \
-    "data": <sensor_data>}, ...]}')
+parser_update.add_argument('-f', '--json-msg-file', type=str, help='JSON File \
+    with update message to be sent. Format: {"id": <device_id>, \
+    "data":[{"sensor_id": <sensor_id>, \
+    "data": <sensor_data>}, ...]}',
+    default='', metavar="MSG_UPDATE_FILE")
```
Comment (ramonhpr) [inline test/mock-connector.py]: `default=json.dumps(defaultMsgUpdate), ...`

File: test/mock-connector.py [hunk 15]
```diff
@@ -59,27 +81,34 @@ def msg_consume(args):
...
 
 parser_request = subparsers.add_parser('send-request', help='Sends a message \
     requesting data from sensor device')
-parser_request.add_argument('message', type=str, help='Request message to be \
-    sent. Format: {"id": <device_id>, "data":[<sensor_id>, ...]}')
+parser_request.add_argument('-f', '--json-msg-file', type=str, help='JSON File \
+    with request message to be sent. Format: {"id": <device_id>, \
+    "data":[{"sensor_id": <sensor_id>, \
+    "data": <sensor_data>}, ...]}',
+    default='', metavar="MSG_REQUEST_FILE")
```
Comment (ramonhpr) [inline test/mock-connector.py]: `default=json.dumps(defaultMsgRequest), ...`

File: test/mock-connector.py (hunk 12 above)
Comment (ramonhpr) [reply test/mock-connector.py]: No need to handle the exception. Let the user know if the file not exists

File: test/mock-connector.py [hunk 16]
```diff
@@ -38,6 +39,27 @@
     level=logging.INFO,
     datefmt='%Y-%m-%d %H:%M:%S')
 
+def __parse_update_message(msg_file):
```
Comment (ramonhpr) [inline test/mock-connector.py]: Remove this function

File: test/mock-connector.py [hunk 17]
```diff
@@ -38,6 +39,27 @@
...
+        with open(msg_file) as fd:
+            msg = json.load(fd)
+    else:
+        msg = {
+            "id": "0123456789abcdef",
+            "data": [{"sensor_id": 253, "data": True}]
+        }
+    return json.dumps(msg)
+
+def __parse_request_message(msg_file):
```
Comment (ramonhpr) [inline test/mock-connector.py]: Remove this function

File: test/mock-connector.py [hunk 18]
```diff
@@ -20,6 +20,7 @@
 
 import pika
 import logging
+import json
```
Comment (ramonhpr) [inline test/mock-connector.py]: About commit message 76787b3d84e4d4291318563074cbdddf2fd8fa07:
In title say that it is related to mock-connector

File: src/msg.c [hunk 19]
```diff
@@ -1403,10 +1403,13 @@ static void start_timeout(struct l_timeout *timeout, void *user_data)
 static void append_on_update_list(void *data, void *user_data)
 {
 	const struct session *session = user_data;
+	knot_msg_schema *schema_found;
```
Comment (ramonhpr) [inline src/msg.c]: About commit message ab14e2e73504ec9050da2978d9a6b50919948c5a:
Verify if sensor id exists in schema in update

This patch verifies if the thing has the sensor id in schema before sending the command update.

Or something like that.

File: test/mock-connector.py [hunk 20]
```diff
@@ -59,27 +81,34 @@ def msg_consume(args):
...
 parser = argparse.ArgumentParser(description='Mock KNoT Fog Connector')
 parser.set_defaults(func=msg_consume)
 subparsers = parser.add_subparsers(help='sub-command help', dest='subcommand')
 
 parser_update = subparsers.add_parser('send-update', help='Sends a message to \
     update the sensor in device')
-parser_update.add_argument('message', type=str, help='Update message to be \
-    sent. Format: {"id": <device_id>, "data":[{"sensor_id": <sensor_id>, \
-    "data": <sensor_data>}, ...]}')
+parser_update.add_argument('-f', '--json-msg-file', type=str, help='JSON File \
```
Comment (ramonhpr) [inline test/mock-connector.py]: I do think that this format string won't handle white spaces.
You can use the parameter `formatter_class` to archive this:https://docs.python.org/3/library/argparse.html#formatter-class

File: test/mock-connector.py (hunk 14 above)
Comment (vitbaq) [reply test/mock-connector.py]: If the default argument is not ' ', I do not have a manner to know if the user passes a file or not.

File: test/mock-connector.py (hunk 15 above)
Comment (vitbaq) [reply test/mock-connector.py]: If the default argument is not ' ', I do not have a manner to know if the user passes a file or not.

File: test/mock-connector.py (hunk 17 above)
Comment (vitbaq) [reply test/mock-connector.py]: Because of the facts already explained, I can not remove.

File: test/mock-connector.py (hunk 16 above)
Comment (vitbaq) [reply test/mock-connector.py]: Because of the facts already explained, I can not remove.

File: test/mock-connector.py (hunk 10 above)
Comment (vitbaq) [reply test/mock-connector.py]: Is not necessary in my opinion, if the function __parse_*_message exist.

File: test/mock-connector.py (hunk 11 above)
Comment (vitbaq) [reply test/mock-connector.py]: Is not necessary in my opinion, if the function __parse_*_message exist.

File: test/mock-connector.py (hunk 12 above)
Comment (vitbaq) [reply test/mock-connector.py]: If I do that, I do not have a manner to know when I send the default message.

File: test/mock-connector.py (hunk 13 above)
Comment (vitbaq) [reply test/mock-connector.py]: If I do that, I do not have a manner to know when I send the default message.

Comment (ramonhpr) [reply test/mock-connector.py]: you are right

**Pull request merged**